modules/automode.cpp

```diff
@@ -0,0 +1,420 @@
+        sLine.Token(3, false, "\t").Split(" ", m_ssChans);
+        return !m_sMode.empty();
+    }
+
+  private:
+  protected:
+    CString m_sUsername;
+    CString m_sMode;
+    set<CString> m_ssHostmasks;
+    set<CString> m_ssChans;
+};
+
+class CAutoModeMod : public CModule {
+  public:
+    MODCONSTRUCTOR(CAutoModeMod) {
+        AddHelpCommand();
+        AddCommand("ListUsers", "", t_d("List all users"),
+                   [=](const CString& sLine) { OnListUsersCommand(sLine); });
+        AddCommand("AddChans", t_d("<user> <channel> [channel] ..."),
+                   t_d("Adds channels to a user"),
+                   [=](const CString& sLine) { OnAddChansCommand(sLine); });
+        AddCommand("DelChans", t_d("<user> <channel> [channel] ..."),
+                   t_d("Removes channels from a user"),
+                   [=](const CString& sLine) { OnDelChansCommand(sLine); });
+        AddCommand("AddMasks", t_d("<user> <mask>,[mask] ..."),
+                   t_d("Adds masks to a user"),
+                   [=](const CString& sLine) { OnAddMasksCommand(sLine); });
+        AddCommand("DelMasks", t_d("<user> <mask>,[mask] ..."),
+                   t_d("Removes masks from a user"),
+                   [=](const CString& sLine) { OnDelMasksCommand(sLine); });
+        AddCommand("AddUser",
+                   t_d("<user> <hostmask>[,<hostmasks>...] <mode> [channels]"),
+                   t_d("Adds a user"),
+                   [=](const CString& sLine) { OnAddUserCommand(sLine); });
+        AddCommand("DelUser", t_d("<user>"), t_d("Removes a user"),
+                   [=](const CString& sLine) { OnDelUserCommand(sLine); });
+    }
+
+    bool OnLoad(const CString& sArgs, CString& sMessage) override {
+        // Load the users
+        for (MCString::iterator it = BeginNV(); it != EndNV(); ++it) {
+            const CString& sLine = it->second;
+            CAutoModeUser* pUser = new CAutoModeUser;
+            if (!pUser->FromString(sLine) ||
+                FindUser(pUser->GetUsername().AsLower())) {
+                delete pUser;
+            } else {
+                m_msUsers[pUser->GetUsername().AsLower()] = pUser;
+            }
+        }
+        return true;
+    }
+
+    ~CAutoModeMod() override {
+        for (const auto& it : m_msUsers) {
+            delete it.second;
+        }
+        m_msUsers.clear();
+    }
+
+    void OnJoin(const CNick& Nick, CChan& Channel) override {
+        // If we have permissions in this chan
+        if (Channel.HasPerm(CChan::Op) || Channel.HasPerm(CChan::HalfOp)) {
+            CheckAutoMode(Nick, Channel);
+        }
+    }
+
+    void OnMode(const CNick* pOpNick, const CNick& Nick, CChan& Channel,
+               bool bNoChange) {
+        if (Nick.GetNick() == GetNetwork()->GetIRCNick().GetNick()) {
+            const map<CString, CNick>& msNicks = Channel.GetNicks();
+            for (const auto& it : msNicks) {
+                if (!it.second.HasPerm(CChan::HalfOp)) {
```

Comment:
Are you sure?

Reply:
Ah, good catch. That needs to become more robust.